### ISSUE TITLE
the interaction of the fireOnOneErrback and consumeErrors parameters to DeferredList is not well documented

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3595]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3595|
|Type|defect|
|Created|2008-12-28 22:19:28Z|
|Branch|https://github.com/twisted/twisted/tree/deferredlist-docs-3595-2|

`fireOnOneErrback` indicates that `DeferredList` will errback when any of the contained Deferreds errbacks.  But `consumeErrors` is documented as meaning that "any errors raised in the original deferreds should be consumed".  This has no effect on `fireOnOneErrback`, but a reader could construe that it does.  Clarifying this might aid readers.

<details><summary>Searchable metadata</summary>

```
trac-id__3595 3595
type__defect defect
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__branches_deferredlist_docs_3595_2 branches/deferredlist-docs-3595-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1230502768000000 1230502768000000
changetime__1278713188000000 1278713188000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Rather than closing this ticket as invalid, I'm going to try to salvage it by changing all of its parameters to that it more usefully described some issue.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [25737]) Branching to 'deferredlist-docs-3595'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

attempt at improved docs in the branch

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @exarkun|
|-|-|

The change in `defer.xhtml` doesn't seem to help much.  There are lots of other things that `consumeErrors` doesn't affect the value of :)

The change to the docstring almost helps, but still doesn't describe the interaction well.  I can't really think off the top of my head of how to explain the interaction myself without a picture of the callback chains, but the important point to communicate is that the failure will be delivered to the callback of the `DeferredList`, but not to any callbacks added _after_ the original `Deferred` is passed to the `DeferredList`'s constructor.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [26062]) Try to clarify the behavior even more

refs [#3595](https://github.com/twisted/twisted/issues/3595)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner|
|-|-|

I left `defer.xhtml` alone.  Maybe the change doesn't help much, but that means it helps a little, right?  Or is the implication that more help should be provided?

I tried some different words in the `DeferredList.__init__` docstring.  Not really sure it's much better.

Another possible solution would be to deprecate `DeferredList` and replace it with simpler things.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_7"></a>|@mwhudson commented|
|-|-|

FWIW, I still don't understand ```DeferredList.__init__``` :(

In particular, I didn't understand what "This L{DeferredList} will still observe L{Failure}s from C{deferredList}." meant until I read this ticket and I'm still not really sure.

I think this ticket does suggest that DeferredList is complicated in not-useful ways.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_8"></a>|@mwhudson set owner to @exarkun|
|-|-|

That wasn't a very good review I guess.

I think probably the only way to be clearer is, in the docstring, enumerate the cases - fireOnOneErrback True and consumeErrors True, then fireOnOneErrback True and consumeErrors False, then ... and explain briefly what happens in each case.  Boring to write and read, but it should at least be clear.

**Comment by twisted-trac** [comment]
|<img alt="cary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|cary set owner to cary|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="cary's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|cary set owner to @exarkun|
|-|-|

DeferredList is going to be deprecated in favor of a simpler API soon.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [29720]) Branching to 'deferredlist-docs-3595-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

I tried again.  I don't know if it's any better this time.

Since the cases _aren't_ actually different, I didn't really want to enumerate them.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_13"></a>|@mwhudson set owner to @exarkun|
|-|-|

The changes now look like an unambiguous improvement to me.

The effect fireOnOne{Errback,Callback} has on the value passed to the callback/errback is not documented anywhere, but that's not new.  If you can't be bothered to address that now, please file a ticket noting this and merge this branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

Expanded the docs to cover all the possible results, in the misreferenced r29759

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [29760]) Merge deferredlist-docs-3595-2

Author: exarkun
Reviewer: glyph, mwh
Fixes: [#3595](https://github.com/twisted/twisted/issues/3595)

Clarify the documentation for the various flags `DeferredList` accepts.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
